packages/plugins/rate-limiter/src/index.ts

```diff
@@ -0,0 +1,53 @@
+import { DefaultContext, Plugin } from '@envelop/types';
+import { IntValueNode, StringValueNode } from 'graphql';
+import { getDirective } from './utils';
+import { getGraphQLRateLimiter } from 'graphql-rate-limit';
+export * from './utils';
+
+export class UnauthenticatedError extends Error {}
+
+export type IdentifyFn<ContextType = unknown> = (context: ContextType) => string;
+
+export const DIRECTIVE_SDL = /* GraphQL */ `
+  directive @rateLimit(max: Int, window: String, message: String) on FIELD_DEFINITION
+`;
+
+export type RateLimiterPluginOptions = {
+  identifyFn: IdentifyFn;
+  rateLimitDirectiveName?: 'rateLimit' | string;
+};
+
+export const useRateLimiter = <ContextType extends DefaultContext = DefaultContext>(
+  options: RateLimiterPluginOptions
+): Plugin<ContextType> => {
+  const rateLimiterFn = getGraphQLRateLimiter({ identifyContext: options.identifyFn });
+  return {
+    async onContextBuilding({ context, extendContext }) {
+      extendContext(({
+        rateLimiterFn,
+      } as unknown) as ContextType);
+    },
+    onExecute() {
+      return {
+        async onResolverCalled({ args, root, context, info }) {
+          const rateLimitDirectiveNode = getDirective(info, options.rateLimitDirectiveName || 'rateLimit');
+
+          if (rateLimitDirectiveNode) {
+            const maxNode = rateLimitDirectiveNode.arguments.find(arg => arg.name.value === 'max').value as IntValueNode;
+            const max = parseInt(maxNode.value);
+            const windowNode = rateLimitDirectiveNode.arguments.find(arg => arg.name.value === 'window').value as StringValueNode;
+            const window = windowNode.value;
+            const messageNode = rateLimitDirectiveNode.arguments.find(arg => arg.name.value === 'message').value as IntValueNode;
+            const message = messageNode.value;
+
+            const errorMessage = await (context as { rateLimiterFn: typeof rateLimiterFn }).rateLimiterFn(
```

Comment:
@kamilkisiela we can improve TS support with generics here, right? I mean to solve all these `unknown` and casts